### PR TITLE
 Make special files (incl. new additional-pip-requirements.txt) configurable

### PR DESCRIPTION
**Pull request by lreiher**
- renames `additional.apt-dependencies` to `additional-debs.txt`
- renames `files` to `additional-files`
- introduces `additional-pip-requirements.txt`
- introduces build arguments for configuring special filenames and toggling recursion (disabled by default)
- fixes support of base images without ROS
- fixes instructions for local build

Closes GitLab issues #39, #28, #44.

### CODE REVIEW

File: .github/workflows/github.yml [hunk 1]
```diff
@@ -23,7 +24,8 @@ jobs:
         with:
           owner: ika-rwth-aachen
           repo: docker-ros-ci
-          ref: main
+          ref: improvement/naming-of-special-files
+          # TODO: restore ref=main
```
Comment (lreiher) [inline .github/workflows/github.yml:28]: TODO: restore ref=main

File: .github/workflows/github.yml [hunk 2]
```diff
@@ -11,7 +11,8 @@ jobs:
...
```
Comment (lreiher) [inline .github/workflows/github.yml:15]: TODO: restore ref=main

File: .github/workflows/gitlab.yml [hunk 3]
```diff
@@ -9,7 +9,8 @@ jobs:
     steps:
       - name: Trigger pipeline
         run: |
-          curl --silent --fail --request POST --form "token=${{ secrets.DOCKER_ROS_CI_TRIGGER_GITLAB_TOKEN }}" --form "ref=main" --form "variables[DOCKER_ROS_GIT_REF]=${{ github.sha }}" "https://gitlab.ika.rwth-aachen.de/api/v4/projects/1886/trigger/pipeline" | jq -r .id > id
+          curl --silent --fail --request POST --form "token=${{ secrets.DOCKER_ROS_CI_TRIGGER_GITLAB_TOKEN }}" --form "ref=improvement/naming-of-special-files" --form "variables[DOCKER_ROS_GIT_REF]=${{ github.sha }}" "https://gitlab.ika.rwth-aachen.de/api/v4/projects/1886/trigger/pipeline" | jq -r .id > id
+      # TODO: restore ref=main
```
Comment (lreiher) [inline .github/workflows/gitlab.yml:13]: TODO: restore ref=main

File: action.yml [hunk 4]
```diff
@@ -65,6 +68,42 @@ inputs:
   git-ssh-known-host-keys:
     description: "Known SSH host keys for cloning private Git repositories via SSH (may be obtained using `ssh-keyscan`)"
 
+  # TODO: where to specify defaults? action.yml + docker-ros.yml vs ci.sh vs dockerfile
```
Comment (lreiher) [inline action.yml:71]: TODO: where to specify defaults? action.yml + docker-ros.yml vs ci.sh vs dockerfile

File: action.yml [hunk 5]
```diff
@@ -65,6 +68,42 @@ inputs:
...
+  # TODO: no default, instead only process if explicitly specified?
+  additional-debs-file:
+    description: "Relative filepath to file containing additional apt deb packages to install"
+    default: docker/additional-debs.txt
+  
+  # TODO: rename to enable_*?
```
Comment (lreiher) [inline action.yml:77]: TODO: rename to enable_*?

File: docker/Dockerfile [hunk 6]
```diff
@@ -82,15 +82,47 @@ RUN echo "set -e" >> $WORKSPACE/.install-dependencies.sh && \
     rm -rf /var/lib/apt/lists/*
 
 # add additionally specified apt dependencies to install script
+# TODO: if repository is top-level package, WORKDIR would need to be $WORKSPACE/src/target/<PACKAGE>
```
Comment (lreiher) [inline docker/Dockerfile:85]: TODO: if repository is top-level package, WORKDIR would need to be $WORKSPACE/src/target/<PACKAGE>

File: docker/Dockerfile [hunk 7]
```diff
@@ -82,15 +82,47 @@ RUN echo "set -e" >> $WORKSPACE/.install-dependencies.sh && \
...
+ARG CUSTOM_SCRIPT_FILE="docker/custom.sh"
+ARG CUSTOM_SCRIPT_RECURSIVE="false"
+RUN if [[ $CUSTOM_SCRIPT_RECURSIVE == 'true' ]]; then \
+        find . -type f -name $(basename ${CUSTOM_SCRIPT_FILE}) -exec sed '$a\' {} >> $WORKSPACE/.install-dependencies.sh \; ; \
+    elif [[ -f ${CUSTOM_SCRIPT_FILE} ]]; then \
+        cat ${CUSTOM_SCRIPT_FILE} >> $WORKSPACE/.install-dependencies.sh ; \
+    fi
 
 # remove now obsolete docker folder from copied repository content to avoid redundancies
+# TODO: there is no need for a docker folder, all could be stored elsewhere
```
Comment (lreiher) [inline docker/Dockerfile:124]: TODO: there is no need for a docker folder, all could be stored elsewhere

File: action.yml (hunk 4 above)
Comment (jpbusch) [reply action.yml:71]: everywhere

Comment (jpbusch) [reply action.yml:71]: also keep defaults for additional special files

File: action.yml (hunk 5 above)
Comment (jpbusch) [reply action.yml:77]: yes

**Review by jpbusch — DISMISSED** [review]

**Review by jpbusch — APPROVED** [review]

File: .gitlab-ci/docker-ros.yml [hunk 8]
```diff
@@ -219,19 +227,19 @@ Test dev-arm64:
...
-      if [[ ${PLATFORM} =~ amd64 && ${PLATFORM} =~ arm64 ]]; then
-        [[ ${TARGET} =~ dev ]] && docker manifest create ${IMG_DEV} --amend ${_IMAGE_DEV_CI_AMD64} --amend ${_IMAGE_DEV_CI_ARM64}
-        [[ ${TARGET} =~ run ]] && docker manifest create ${IMG_RUN} --amend ${_IMAGE_RUN_CI_AMD64} --amend ${_IMAGE_RUN_CI_ARM64}
-      elif [[ ${PLATFORM} == amd64 ]]; then
-        [[ ${TARGET} =~ dev ]] && docker manifest create ${IMG_DEV} --amend ${_IMAGE_DEV_CI_AMD64}
-        [[ ${TARGET} =~ run ]] && docker manifest create ${IMG_RUN} --amend ${_IMAGE_RUN_CI_AMD64}
-      elif [[ ${PLATFORM} == arm64 ]]; then
-        [[ ${TARGET} =~ dev ]] && docker manifest create ${IMG_DEV} --amend ${_IMAGE_DEV_CI_ARM64}
-        [[ ${TARGET} =~ run ]] && docker manifest create ${IMG_RUN} --amend ${_IMAGE_RUN_CI_ARM64}
+      if [[ "${PLATFORM}" =~ amd64 && "${PLATFORM}" =~ arm64 ]]; then
```
Comment (lreiher) [inline .gitlab-ci/docker-ros.yml:230]: Why is the quoting necessary? Do we need to quote amd64/arm64 as well?

File: docker/Dockerfile [hunk 9]
```diff
@@ -155,6 +162,21 @@ RUN apt-get update && \
     $WORKSPACE/.install-dependencies.sh && \
     rm -rf /var/lib/apt/lists/*
 
+# install essential build and ROS CLI tools
+RUN apt-get update && \
+    apt-get install -y \
+        build-essential \
```
Comment (lreiher) [inline docker/Dockerfile:168]: Move build-essential before install script execution?

File: README.md [hunk 10]
```diff
@@ -21,7 +21,8 @@
   - [Build multi-arch images on arch-specific self-hosted runners in parallel](#build-multi-arch-images-on-arch-specific-self-hosted-runners-in-parallel)
   - [Build images locally](#build-images-locally)
 - [Advanced Dependencies](#advanced-dependencies)
-  - [Extra System Dependencies (apt)](#extra-system-dependencies-apt)
+  - [Extra System Dependencies (*apt*)](#extra-system-dependencies-apt)
+  - [Extra System Dependencies (*pip*)](#extra-system-dependencies-pip)
   - [Custom Installation Script](#custom-installation-script)
   - [Extra Image Files](#extra-image-files)
 - [Configuration Variables](#configuration-variables)
```
Comment (lreiher) [inline README.md:28]: update version number to 1.1.0 in readme (and search for any other occurrences)

Comment (jpbusch) [reply README.md:28]: Especially [this](https://github.com/ika-rwth-aachen/docker-ros/tree/improvement/naming-of-special-files#about) section needs to be updated

File: .gitlab-ci/docker-ros.yml (hunk 8 above)
Comment (jpbusch) [reply .gitlab-ci/docker-ros.yml:230]: Its just more robust and can handle e.g. a space in the variable (`PLATFORM=dev, run`)

**Review by jpbusch — APPROVED** [review]

**Pull request merged**